src/dataMap/metaManager/lazyFactoryMap.js

```diff
@@ -0,0 +1,315 @@
+ *    │       │      │      ┌─────────────┘│    ││
+ *    │       │      │      │      ┌───────┼────┘│
+ *    │       │      │      │      │       │     │
+ * +------+------+------+------+------+------+------+
+ * | AAA  | BBB  | CCC  | DDD  | EEE  | FFF  | GGG  |  Data
+ * +------+------+------+------+------+------+------+
+ *
+ * +------+------+------+
+ * | 0/10 | 1/13 | 2/14 |  Keys after
+ * +------+------+------+
+ *    │       │      │
+ *    │       │      └─────────────┐
+ *    │       └────────────┐       │
+ *    │                    │       │
+ *    │                    │       │
+ *    │                    │       │
+ *    │                    │       │
+ * +------+------+------+------+------+------+------+
+ * | AAA  | BBB  | CCC  | DDD  | EEE  | FFF  | GGG  |  Data
+ * +------+------+------+------+------+------+------+
+ *           /│\   /│\                   /│\   /│\
+ *            └──┬──┘                     └──┬──┘
+ *           This data is marked as "hole" which
+ *           means that can be replaced by new item
+ *           when that will be created.
+ *
+ * map.obtain(2) // returns the value ("EEE") as it should. Access to the value is
+ *               // changed (the key was downshifted). However, the internal index has not changed,
+ *               // which means that the data does not need to be changed (spliced) too.
+ *
+ * After previous remove operation which creates some "holes" obtaining new
+ * items replaces that "holes" as follows:
+ *
+ * // Obtains new item
+ * map.obtain(90) // Returns "NEW" value
+ *
+ * +------+------+------+...+------+
+ * | 0/10 | 1/13 | 2/14 |   | 90/0 |  Keys after
+ * +------+------+------+...+------+
+ *    │       │      │          │
+ *    │       │      └──────────┼────────────┐
+ *    │       └─────────────────┼─────┐      │
+ *    └──────────┐              │     │      │
+ *               │              │     │      │
+ *    ┌──────────┼──────────────┘     │      │
+ *    │          │                    │      │
+ * +------+...+------+------+------+------+------+-----+
+ * | NEW  |   | AAA  | BBB  | CCC  | DDD  | EEE  | FFF |  Data
+ * +------+...+------+------+------+------+------+-----+
+ *   /│\
+ *    │
+ * The first "hole" (at index 0) item is permanently removed and replaced by a new item.
+ * The hole index is taken from the hole collection which act as FIFO (First In First Out).
+ */
+export default class LazyFactoryMap {
+  constructor(valueFactory) {
+    this.valueFactory = valueFactory;
+    /**
+     * An array which contains data.
+     *
+     * @type {Array}
+     */
+    this.data = [];
+    /**
+     * An array of indexes where the key of the array is mapped to the value which points to the
+     * specific position of the data array.
+     *
+     * @type {Number[]}
+     */
+    this.index = [];
+    /**
+     * The collection of indexes where points to the data items which can be replaced by obtaining new
```

Comment:
The collection of indexes that points..? :)

Reply:
Fixed within https://github.com/handsontable/handsontable/pull/6254/commits/0d440ef70170a376b2ffa1b3aa0ec2f329464947.